### PR TITLE
Add generator message with info on how to secure a function

### LINKED ISSUE #2014: Add message about security to generated custom functions
The function generator and template JSDoc will link to RedwoodJS.com documentation to help the develop understand how to better secure their function when deployed, if needed and appropriate, since many may not be aware that when deployed, the function is an open api endpoint that anyone can access.

See: https://github.com/redwoodjs/redwoodjs.com/pull/659

### PR DESCRIPTION
**Pull request by dthyresson**
Fixes #2014 

The function generator and template JSDoc links to documentation to help the develop understand how to better secure their function when deployed, if needed and appropriate, since many may not be aware that when deployed, the function is an open api endpoint that anyone can access.

Also see: https://github.com/redwoodjs/redwoodjs.com/pull/659 in RedwoodJS.com repo

### CODE REVIEW

File: packages/cli/src/commands/generate/function/function.js [hunk 1]
```diff
@@ -54,13 +53,44 @@ export const builder = (yargs) => {
...
   Object.entries(yargsDefaults).forEach(([option, config]) => {
     yargs.option(option, config)
   })
 }
 
-export const { command, handler } = createYargsForComponentGeneration({
-  componentName: 'function',
-  filesFn: files,
-})
+export const handler = async ({ force = false }) => {
```
Comment (dthyresson) [inline packages/cli/src/commands/generate/function/function.js:63]: Is this correct to create my own handler - so I can console log some messaging after the tasks run?

File: packages/cli/src/commands/generate/function/__tests__/fixtures/jsFunc.js [hunk 2]
```diff
@@ -1,18 +1,34 @@
+import { logger } from 'src/lib/logger'
```
Comment (dthyresson) [inline packages/cli/src/commands/generate/function/__tests__/fixtures/jsFunc.js:1]: Am adding logger to enforce best practices since on of the security considerations will be visibility and one should log what the function is doing to see if an abuse.

File: packages/cli/src/commands/generate/function/__tests__/fixtures/jsFunc.js [hunk 3]
```diff
@@ -1,18 +1,34 @@
...
+    return {
+      statusCode: 200,
+      body: JSON.stringify({
+        data: 'javascriptFunction function',
+      }),
+    }
+  } catch (error) {
+    logger.error({ error }, error.message)
+
+    return {
```
Comment (dthyresson) [inline packages/cli/src/commands/generate/function/__tests__/fixtures/jsFunc.js:30]: Want to show that a 401 unauthorized is appropriate in some cases.

When the `secureHandler` work is done in another PR that provides verification methods can leverage that here.

File: packages/cli/src/commands/generate/function/function.js (hunk 1 above)
Comment (dthyresson) [reply packages/cli/src/commands/generate/function/function.js:63]: Looks like it is, I just was not setting the command and some params approriately.

image

**Comment by dthyresson** [comment]
Improved formatting of message. Open to copy suggestions.

Screen Shot 2021-04-04 at 09 51 03

Note the link to the docs is TBD in the redwoodjs.com repo as draft.

**Comment by dthyresson** [comment]
Note: the link seems to have some extra encoded chars on it

```
https://redwoodjs.com/docs/serverless-functions#security-considerations%E2%80%8B
````

But I think this is a VSCode terminal issue as different terminals render links differently, see iTerm here:

image

And this issue in terminal-link:

https://github.com/sindresorhus/terminal-link/issues/11

**Comment by dthyresson** [comment]
Also, confirmed generated function still works given

```terminal
yarn rw g function publicFunction --force
```

image

File: packages/cli/src/commands/generate/function/__tests__/fixtures/jsFunc.js (hunk 3 above)
Comment (peterp) [reply packages/cli/src/commands/generate/function/__tests__/fixtures/jsFunc.js:30]: The try/ catch part is a bit strange for me, since the return statement can't fail. Maybe a comment feel less strange until the secure handler work is done?

Comment (dthyresson) [reply packages/cli/src/commands/generate/function/__tests__/fixtures/jsFunc.js:30]: > Maybe a comment feel less strange until the secure handler work is done?

Agreed. I'll remove until I can hook in `secureHandler`.

I was actually thinking `secureHandler` *could* have a verification method like `public` or `open` that always returns true so that that could be baked into the function template.

Then one just swaps out the verification method to SHA256 or whatnot.

Comment (dthyresson) [reply packages/cli/src/commands/generate/function/__tests__/fixtures/jsFunc.js]: @peterp I couldn't think of a good way to comment the 401 case, so I simply removed the try/catch example. but I left in a small log statement as wanted to get logger into the template.

Once secureHandler is done I'll revisit this template.

File: packages/cli/src/commands/generate/function/function.js [hunk 4]
```diff
@@ -48,19 +48,65 @@ export const builder = (yargs) => {
       description: 'Name of the Function',
       type: 'string',
     })
+    .option('force', {
```
Comment (thedavidprice) [inline packages/cli/src/commands/generate/function/function.js:51]: This PR seems out-of-sync with the recent conversion to TS, which includes the options for either JS or TS language target. See:
https://github.com/redwoodjs/redwood/pull/1854/files#diff-9434dff1ad2a320fdf2411f1471ac4bc3a5b8202d59c8b4f5ac3bcd51b0c26a4

Do you just need to update from main?

File: packages/cli/src/commands/generate/function/function.js [hunk 5]
```diff
@@ -48,19 +48,65 @@ export const builder = (yargs) => {
...
+  try {
+    await tasks.run()
+
+    console.info('')
+
+    console.info(c.warning('⚠ Important:'))
+    console.info('')
+    console.info(
+      c.bold(
+        'When deployed, a custom serverless function is an open API endpoint and is your responsibility to secure appropriately.'
```
Comment (thedavidprice) [inline packages/cli/src/commands/generate/function/function.js:93]:
```suggestion
        'When deployed, a custom serverless function is an open API endpoint ' +
          'and is your responsibility to secure appropriately.'
```

line-break'n

File: packages/cli/src/commands/generate/function/function.js [hunk 6]
```diff
@@ -48,19 +48,65 @@ export const builder = (yargs) => {
...
+    )
+
+    console.info('')
+
+    console.info(
+      c.info(
+        `Please consult the ${terminalLink(
+          'Serverless Function Considerations',
+          'https://redwoodjs.com/docs/serverless-functions#security-considerations'
+        )} in the RedwoodJS documentation for more information.`
```
Comment (thedavidprice) [inline packages/cli/src/commands/generate/function/function.js:104]:
```suggestion
        `Please consult the following documentation for information:\n${terminalLink(
          'Serverless Function Considerations',
          'https://redwoodjs.com/docs/serverless-functions#security-considerations'
        )}`
```

The link was getting a bit lost in the text for me. What do you think about this? (Also, the 'info' color is super light in my console... not sure there's much to do about that in this PR.)

Screen Shot 2021-04-05 at 11 13 28 PM

**Review by thedavidprice — CHANGES_REQUESTED** [review]
This is turning out great @dthyresson And definitely worth the time you've spent to get the details right.

I'll keep a look out for the doc PR to align with v0.29

File: packages/cli/src/commands/generate/function/function.js (hunk 6 above)
Comment (dthyresson) [reply packages/cli/src/commands/generate/function/function.js:101]: Added line break. Done.

> light in my console... not sure there's much to do about that in this PR.

That's VSCode's gray and what "info" is set to in colors.

But am changing to not use info and looks better:

image

File: packages/cli/src/commands/generate/function/function.js (hunk 4 above)
Comment (dthyresson) [reply packages/cli/src/commands/generate/function/function.js]: I had added back the "--force" option manually as I didn't realize it was already included. The rest of the file is the TS flavor/latest.

File: packages/cli/src/commands/generate/function/function.js [hunk 7]
```diff
@@ -54,13 +54,53 @@ export const builder = (yargs) => {
...
   })
 }
 
-export const { command, handler } = createYargsForComponentGeneration({
-  componentName: 'function',
-  filesFn: files,
-})
+// This could be built using createYargsForComponentGeneration;
+// however, we need to add a message after generating the function files
+export const handler = async ({ name, force }) => {
```
Comment (thedavidprice) [inline packages/cli/src/commands/generate/function/function.js:66]: This needs to take the --javascript and --typescript options, defaulting to javascript.

Currently, the command _only_ generates .ts regardless

Comment (thedavidprice) [reply packages/cli/src/commands/generate/function/function.js:66]: Higher-level question not to be resolved in this PR --> for our generator commands in general, do we have unit tests confirming the conversion to JS is working? AND that TS files can be generated as well. 🤔

Comment (dthyresson) [reply packages/cli/src/commands/generate/function/function.js:66]: Oh - when I read the comment in 

```js
  // Add default options, includes '--typescript', '--javascript', '--force', ...
  Object.entries(yargsDefaults).forEach(([option, config]) => {
    yargs.option(option, config)
  })
```
I assumed it did that -- but apparently it doesn't. Well it did add --force so I imagine the --javascript is added too but just not handled.

Comment (dthyresson) [reply packages/cli/src/commands/generate/function/function.js:66]: > for our generator commands in general, do we have unit tests confirming the conversion to JS is working?

Yes, from a templating standpoint:

```js
  javascriptFiles = functionGenerator.files({
    name: 'javascript-function',
    javascript: true,
  })
  typescriptFiles = functionGenerator.files({
    name: 'typescript-function',
    typescript: true,
  })
```

The issue here is that it sets the `javascript` or `typescript` but in the function generator, that param wasn't being passed to the `files()`.

Comment (dthyresson) [reply packages/cli/src/commands/generate/function/function.js:66]: > Currently, the command _only_ generates .ts regardless

Fixed:

```terminal

rw_tests % yarn rw g function myTypescriptFunction --typescript 
yarn run v1.22.10
$ '/rw_tests/node_modules/.bin/rw' g function myTypescriptFunction --typescript
  ✔ Generating function files...
    ✔ Successfully wrote file `./api/src/functions/myTypescriptFunction.ts`

⚠ Important:

When deployed, a custom serverless function is an open API endpoint and is your responsibility to secure appropriately.

Please consult the Serverless Function Considerations (​https://redwoodjs.com/docs/serverless-functions#security-considerations​) in the RedwoodJS documentation for more information.

✨  Done in 2.10s.
rw_tests % yarn rw g function myJavascriptFunction --javascript 
yarn run v1.22.10
$ '/rw_tests/node_modules/.bin/rw' g function myJavascriptFunction --javascript
  ✔ Generating function files...
    ✔ Successfully wrote file `./api/src/functions/myJavascriptFunction.js`

⚠ Important:

When deployed, a custom serverless function is an open API endpoint and is your responsibility to secure appropriately.

Please consult the Serverless Function Considerations (​https://redwoodjs.com/docs/serverless-functions#security-considerations​) in the RedwoodJS documentation for more information.

✨  Done in 2.41s.
rw_tests % yarn rw g function myDefaultJavascriptFunction              
yarn run v1.22.10
$ '/rw_tests/node_modules/.bin/rw' g function myDefaultJavascriptFunction
  ✔ Generating function files...
    ✔ Successfully wrote file `./api/src/functions/myDefaultJavascriptFunction.js`

⚠ Important:

When deployed, a custom serverless function is an open API endpoint and is your responsibility to secure appropriately.

Please consult the Serverless Function Considerations (​https://redwoodjs.com/docs/serverless-functions#security-considerations​) in the RedwoodJS documentation for more information.

✨  Done in 2.30s.
```

**Comment by dthyresson** [comment]
@thedavidprice Testing the function generator actually creating ts or js from command line might be an e2e test vs the unit test that exists currently.

**Review by thedavidprice — APPROVED** [review]

**Comment by thedavidprice** [comment]
Thanks DT!! I'll keep a look out for docs.
